services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy

```diff
@@ -0,0 +1,942 @@
+        flowHelper.deleteFlow(flow.id)
+        database.resetCosts()
+
+        where:
+        flowDescription | bandwidth
+        "a metered"     | 1000
+        "an unmetered"  | 0
+    }
+
+    @Unroll
+    def "Able to get flow(protectedPath=#protectedPath) path in case flow is not rerouted"() {
+        given: "Two active neighboring switches and a flow"
+        def isls = topology.getIslsForActiveSwitches()
+        def (srcSwitch, dstSwitch) = [isls.first().srcSwitch, isls.first().dstSwitch]
+
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow.allocateProtectedPath = protectedPath
+        flowHelper.addFlow(flow)
+
+        when: "Init reroute"
+        def rerouteResponse = northbound.rerouteFlow(flow.id)
+
+        then: "Flow is not rerouted"
+        !rerouteResponse.rerouted
+
+        and: "System allows to get flow path"
+        northbound.getFlowPath(flow.id)
+
+        and: "Cleanup: revert system to original state"
+        flowHelper.deleteFlow(flow.id)
+
+        where:
+        protectedPath << [false, true]
+    }
+
+    def "Unable to create a flow with the 'protected path' option in case there is not enough bandwidth"() {
+        given: "Two active neighboring switches"
+        def isls = topology.getIslsForActiveSwitches()
+        def (srcSwitch, dstSwitch) = [isls.first().srcSwitch, isls.first().dstSwitch]
+
+        and: "Update all isls which can be used by protected path"
+        def bandwidth = 100
+        isls[1..-1].each { database.updateIslAvailableBandwidth(it, 90) }
+
+        when: "Create flow with protected path"
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow.maximumBandwidth = bandwidth
+        flow.allocateProtectedPath = true
+        flowHelper.addFlow(flow)
+
+        then: "Human readable error is returned"
+        def exc = thrown(HttpClientErrorException)
+        exc.rawStatusCode == 404
+        exc.responseBodyAsString.to(MessageError).errorMessage ==
+                "Could not create flow: Not enough bandwidth found or path not found. " +
+                "Couldn't find non overlapping protected path"
+
+        and: "Cleanup: restore available bandwidth"
+        isls.each { database.resetIslBandwidth(it) }
+    }
+
+    def "Unable to update the 'protected path' option on a flow in case there is not enough bandwidth"() {
+        given: "Two active neighboring switches"
+        def isls = topology.getIslsForActiveSwitches()
+        def (srcSwitch, dstSwitch) = [isls.first().srcSwitch, isls.first().dstSwitch]
+
+        and: "Update all isls which can be used by protected path"
+        def bandwidth = 100
+        isls[1..-1].each { database.updateIslAvailableBandwidth(it, 90) }
+
+        when: "Create flow without protected path"
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow.maximumBandwidth = bandwidth
+        flow.allocateProtectedPath = false
+        flowHelper.addFlow(flow)
```

Comment:
In case of parallel links between switches, as far as I understand we may have an error. The flow path may be put on the second link where bandwidth is small.

Reply:
Please, ignore this comment. I had a wrong assumption on it.